server/src/main/java/org/elasticsearch/index/store/AsyncDirectIOIndexInput.java

```diff
@@ -0,0 +1,539 @@
+                final int slot;
+                Integer existingSlot = this.posToSlot.get(pos);
+                if (existingSlot != null && prefetchThreads.get(existingSlot) != null) {
+                    // already being prefetched and hasn't been consumed.
+                    // return early
+                    return;
+                }
+                if (this.posToSlot.size() < maxConcurrentPrefetches && slots.isEmpty() == false) {
+                    slot = slots.removeFirst();
+                    posToSlot.put(pos, slot);
+                    prefetchPos[slot] = pos;
+                } else {
+                    slot = -1;
+                    pendingPrefetches.addLast(pos);
+                }
+                if (slot != -1) {
+                    startPrefetch(pos, slot);
+                }
+                pos += prefetchBytesSize;
+                numSlots--;
+            }
+
+        }
+
+        /**
+         * Try to read the requested bytes from an already prefetched buffer.
+         * If the requested bytes are not in a prefetched buffer, return false.
+         * @param pos the absolute position to read from
+         * @param slice the buffer to read into, must be pre-sized to the required length
+         * @param delta an offset into the slice buffer to start writing at
+         * @return true if the requested bytes were read from a prefetched buffer, false otherwise
+         * @throws IOException if an I/O error occurs
+         */
+        boolean readBytes(long pos, ByteBuffer slice, int delta) throws IOException {
+            final var entry = this.posToSlot.floorEntry(pos + delta);
+            if (entry == null) {
+                return false;
+            }
+            final int slot = entry.getValue();
+            final long prefetchedPos = entry.getKey();
+            // determine if the requested pos is within the prefetched range
+            if (pos + delta >= prefetchedPos + prefetchBytesSize) {
+                return false;
+            }
+            final Future<ByteBuffer> thread = prefetchThreads.get(slot);
+            if (thread == null) {
+                // free slot and decrement active prefetches
+                clearSlotAndMaybeStartPending(slot);
+                return false;
+            }
+            final ByteBuffer prefetchBuffer;
+            try {
+                prefetchBuffer = thread.get();
+            } catch (ExecutionException e) {
+                clearSlotAndMaybeStartPending(slot);
+                IOException ioException = (IOException) ExceptionsHelper.unwrap(e, IOException.class);
+                if (ioException != null) {
+                    throw ioException;
+                }
+                throw new IOException(e.getCause());
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                return false;
```

Comment:
Is the caller expected to retry? This doesn't clear the slot.

More generally, a slot is only cleared when it's read, which means that if a prefetch'd position is never actually read (say, an exception is thrown elsewhere), the results just sit there taking up a slot until the input is closed. Do we need some kind of timeout/clearup mechanism?

Reply:
> Is the caller expected to retry? This doesn't clear the slot.

Hmmmm, yeah, it should likely clear. But if this main thread got an interrupt, doesn't that mean ES is getting shut down?

Reply:
Not necessarily, although Elasticsearch doesnt really use interrupts. It can probably be left as-is then - although that still leaves the general problem of un-read prefetches blocking slots.

Reply:
I will clean this up.